mne/stats/regression.py

```diff
@@ -134,3 +134,163 @@ def _fit_lm(data, design_matrix, names):
+    tstep : None | float
+        If reject is not None, length of windows for peak-to-peak detection.
+
+    Returns
+    -------
+    ev_dict : dictionary
+        A dictionary where the keys correspond to conditions and the values are
+        Evoked objects with the rE[R/F]Ps. These can be used exactly like any
+        other Evoked object, including e.g. plotting or statistics.
+    """
+    sfreq = raw.info["sfreq"]
+    data, times = raw[:]
+    events = events.copy()
+    events[:, 0] -= raw.first_samp
+    conds = list(event_id.keys())
+    if covariates is not None:
+        conds += list(covariates.keys())
+
+    if isinstance(tmin, (float, int)):
+        tmin = dict((cond, int(tmin * sfreq)) for cond in conds)
+    else:
+        tmin = dict((cond, int(tmin.get(cond, -.1) * sfreq))
+                    for cond in conds)
+    if isinstance(tmax, (float, int)):
+        tmax = dict((cond, int(tmax * sfreq) + 1) for cond in conds)
+    else:
+        tmax = dict((cond, int(tmax.get(cond, 1) * sfreq) + 1)
+                    for cond in conds)
+
+    cond_length = dict()
+    # construct predictor matrix
+    # !!! this should probably be improved (including making it more robust to
+    # high frequency data) by operating on sparse matrices. As-is, high
+    # sampling rates plus long windows blow up the system due to the inversion
+    # of massive matrices.
+    # Furthermore, assigning to a preallocated array would be faster.
+    pred_arrays = list()
+    for cond in conds:
+        tmin_, tmax_ = tmin[cond], tmax[cond]
+        n_lags = tmax_ - tmin_
+        # create the first row and column to be later used by toeplitz to build
+        # the full predictor matrix
+        samples, lags = np.zeros(len(times)), np.zeros(n_lags)
+        if cond in event_id.keys():  # for binary predictors
+            ids = ([event_id[cond]] if isinstance(event_id[cond], int)
+                   else event_id[cond])
+            samples[np.asarray([t for t, _, e in events
+                                if e in ids]) + tmin_] = 1  # np.in1d
+        else:  # for predictors from covariates, e.g. continuous ones
+            if len(covariates[cond]) != len(events):
+                error = """Condition {} from ```covariates``` is
+                        not the same length as ```events```""".format(cond)
+                raise ValueError(error)
+            for tx, v in zip(events[:, 0], covariates[cond]):
+                samples[tx + tmin_] = np.float(v)
+        cond_length[cond] = len(np.nonzero(samples))
+
+        # this is the magical part (thanks to Marijn van Vliet):
+        # use toeplitz to construct series of diagonals
+        pred_arrays.append(linalg.toeplitz(samples, lags))
+
+    big_arr = np.hstack(pred_arrays).T
+    # find only those positions where at least one predictor isn't 0
+    has_val = np.asarray([len(np.nonzero(line)[0]) > 0
+                          for line in big_arr.T])
```

Comment:
common, this is not beautifyl and can be vecotized easily, no? Isn't it just np.nonzero on big_array?